components/scream/src/share/util/scream_kokkos_utils.hpp

```diff
@@ -117,61 +119,135 @@ struct ExeSpaceUtils<Kokkos::Cuda> {
+ * Specialization for OpenMP execution space
+ */
 #ifdef KOKKOS_ENABLE_OPENMP
-    return omp_get_thread_num() / _team_size;
-#else
-    return 0;
-#endif
-  }
+template <>
+class TeamUtils<Kokkos::OpenMP> : public _TeamUtilsCommonBase<Kokkos::OpenMP>
+{
+ public:
+  template <typename TeamPolicy>
+  TeamUtils(const TeamPolicy& policy) : _TeamUtilsCommonBase<Kokkos::OpenMP>(policy)
+  { }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& team_member) const
+  { return omp_get_thread_num() / _team_size; }
 };
+#endif
 
 /*
- * Specialization for Cuda execution space. On GPU, we expect all teams to run
- * at once.
+ * Specialization for Cuda execution space.
  */
 #ifdef KOKKOS_ENABLE_CUDA
 template <>
-class TeamUtils<Kokkos::Cuda>
+class TeamUtils<Kokkos::Cuda> : public _TeamUtilsCommonBase<Kokkos::Cuda>
 {
-  int _num_teams;
+  using Device = Kokkos::Device<Kokkos::Cuda, typename Kokkos::Cuda::memory_space>;
+  using flag_type = int; // this appears to be the smallest type that correctly handles atomic operations
+  using view_1d = typename KokkosTypes<Device>::view_1d<flag_type>;
 
-public:
+  view_1d _open_ws_slots;
+
+ public:
   template <typename TeamPolicy>
-  TeamUtils(const TeamPolicy& policy) { _num_teams = policy.league_size(); }
+  TeamUtils(const TeamPolicy& policy) :
+    _TeamUtilsCommonBase<Kokkos::Cuda>(policy),
+    _open_ws_slots("open_ws_slots", _num_teams)
+  { }
 
-  int get_num_concurrent_teams() const { return _num_teams; }
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& team_member) const
+  {
+    if (team_member.league_size() <= _num_teams) {
+      return team_member.league_rank();
+    }
+    else {
+      int ws_idx = 0;
+      Kokkos::single(Kokkos::PerTeam(team_member), [&] () {
+        ws_idx = team_member.league_rank() % _num_teams;
+        while (!Kokkos::atomic_compare_exchange_strong(&_open_ws_slots(ws_idx), (flag_type) 0, (flag_type)1)) {
+          // or random?
+          ws_idx = (ws_idx+1) % _num_teams;
+        }
+      });
+
+      // broadcast the idx to the team with a simple reduce
+      int ws_idx_max_reduce;
+      Kokkos::parallel_reduce(Kokkos::TeamThreadRange(team_member, 1), [&] (int, int& ws_idx_max) {
+        ws_idx_max = ws_idx;
+      }, Kokkos::Max<int>(ws_idx_max_reduce));
+      team_member.team_barrier();
+      return ws_idx_max_reduce;
+    }
+  }
 
   template <typename MemberType>
   KOKKOS_INLINE_FUNCTION
-  int get_workspace_idx(const MemberType& team_member) const { return team_member.league_rank(); }
+  void release_workspace_idx(const MemberType& team_member, int ws_idx) const
+  {
+    if (team_member.league_size() > _num_teams) {
+      team_member.team_barrier();
+      Kokkos::single(Kokkos::PerTeam(team_member), [&] () {
+        flag_type volatile* const e = &_open_ws_slots(ws_idx);
+        *e = (flag_type)0;
+      });
+      team_member.team_barrier();
```

Comment:
I think this second barrier is unnecessary; only the first, at line 243, is necessary.

Reply:
Good catch; this line was a leftover from when I was desperately throwing barriers everywhere, trying to figure out why my test wasn't working (the one you helped me fix by adding `volatile`).